### DISCUSSION TITLE
Price promotions (discounts)

### DISCUSSION THREAD

**Discussion by raphodn** [openfoodfacts/open-prices#508] [General]
This discussion will help gather information on price promotions.
And find a quick way to better manage them in the backend & frontend.

# List of promotion types

|Name|Example|Details|Status|
|---|---|---|---|
|Discount on a product||||
|Seasonal discount|e.g. clothing|||
|Discount on a product because expiring soon||||
|Loyalty card||Discount only if you are a member. Can be immediate, or money saved in your loyalty account||
|Buy X get Y free (BOGO)|Buy 2 get 1 free|||
|Discount above X quantity|Buy 1 get the 2nd at 50%|||
|Pick it yourself|- 5% discount on books<br>- Discount if picked at the farm||
|Second hand|e.g. books, clothing||
|||||

# Related issues

- https://github.com/openfoodfacts/open-prices/issues/76 (current implementation)
- https://github.com/openfoodfacts/open-prices/issues/495
- https://github.com/openfoodfacts/open-prices-frontend/issues/537
- https://github.com/openfoodfacts/open-prices-frontend/issues/840
- https://github.com/openfoodfacts/open-prices/issues/472

# Implementations

## v1

3 fields: 
- `price`
- `price_is_discounted` (boolean)
- `price_without_discount`

## v2

a 4th field was added:
- `discount_type`: quantity, sale, seasonal, loyalty program, expires soon, pick it yourself, other

**Comment by serpico** [comment]
in the case of [#840](https://github.com/openfoodfacts/open-prices-frontend/issues/840) I was also hoping/trying to think of a way to warn user for this specific case, it says discount on the price tag but the price is the same than it was for the last X weeks, so there's no real discount ( I admit I don't know if it's forbidden, to advertise a discount while nothing changed ). I thought about a special color on the graph for that the price points of discounted price ( since the graph will be flat we could see the discount is bogus - and the opposite is true ) ; we could also use some stat i.e mean/average price for the last month, last 2 months in order to give an indication of the extent of the discount.

**Comment by serpico** [comment]
funny thing, when there's 2 prices ( normal price and discounted price ) for the same day/same location as for https://prices.openfoodfacts.org/products/3564706596788?display=chart it's plotted on the graph as a single price point which seems to be the average of the two
![bug average of 2 prices in case of discount_crop](https://github.com/user-attachments/assets/ec719da1-bb1c-46d4-8a16-8972899a3ac6)

 see also [302](https://github.com/openfoodfacts/open-prices-frontend/issues/302)

**Reply by raphodn** [reply]
indeed, currently the chart's Y axis does a "mean aggregation" of the prices for a given day
here : https://github.com/openfoodfacts/open-prices-frontend/blob/646ca0ff9f98d3e2eb84cb281097702402aab78f/src/components/PriceChart.vue#L48

**Reply by tradmangh** [reply]
> funny thing, when there's 2 prices ( normal price and discounted price ) for the same day/same location it's plotted on the graph as a single price point which seems to be the average of the two 

not funny at all ;) and it shows that for complex data there is no such thing as a "simple chart" ;)

different prices ...
 same location: _not ok_, data quality issue
 different location, same store chain: _probably not ok_, _depends_
 different location & store chain: _should be visualized as Min/Max/Avg for each datapoint_, which makes more sense
 ...

![image](https://github.com/user-attachments/assets/ad95e4c5-baac-4e8b-a174-00baa93be14b)

**Comment by tradmangh** [comment]
> Discount on a product because expiring soon

"to good to go" discounts are a whole different story, because they only apply to specific physical product items. While that information might be interesting for certain parties, for useful stats it would need a huge amount of datapoints and regular data capture. Also very rare and short lived.

> Discount above quantity x	Buy 1 get the 2nd at 50%

For food products the most common case is a bulk discount, x% discount on all items of the same if you buy more than x

Other variations for discounts:

- "Member_only" which are often combined with BOGO 
- "x% on the most/least expensive item" but I have not seen it with food products so far

**Comment by serpico** [comment]
to add to the recap table :

Some [shop=farm](https://wiki.openstreetmap.org/wiki/Tag:shop%3Dfarm) offer a "pick it yourself" discount, you can either buy already harvested produce in a display manned by an attendant/cashier but you can also pick/harvest the produce yourself in the fields and they're discounted. Prices for picking yourself are displayed on a big whiteboard/chalkboard.

Those cases could also be interesting candidates for crowdsourcing the extraction of prices from pictures ( https://github.com/openfoodfacts/open-prices/discussions/523 ) as you need 3 to 4 pictures to catch the whole board and it's evolving depending on the harvest season - those kind of shops generally have a limited presence online and from a customer's perspective you can't really plan on buying "to harvest" and generally have to wait till you're in front of the board to know current prices and availability.

That's one of the cases where I feel open prices could be a killer app - but a special tag as envisioned by @TTalex in #523 might be necessary to ease collaboration for an informal group of customers of the same shop, sharing the workload and keeping the prices updated - it could work great for a _projet du mois_-like ( https://github.com/openfoodfacts/open-prices-frontend/issues/934 again brought up by @TTalex  ) using overpass turbo query https://overpass-turbo.eu/s/1Te8. to identify those shops...

Different cases of prices I've seen in those shops :

- Price per unit or kg at the display - depending on produce there's sometimes a discount depending on the amount. 
- Discounted prices at the display for _légumes moches_ , sold per unit or per kg.
- Prepackaged/preweighted amount ( net of 5/10kg for potatoes, onions ).
- Price per unit or kg if the customer pick the produce herself/himself in the field, usually it's a flat price per kg/unit regardless if you take an overflowing wheelbarrow of produce or not....
- Discount for big amount to be picked by the customer ( end of season for tomatoes, I've seen discount if the customer pick and buy 10 kg in one go ).

I feel you would need up to as many pictures as there's cases, in a recent example for yellow onions : 
- 1 picture with prices at the display for onions at the display ( 2 different prices whether it's per 1 kg or per 2 kg ).
- 1 picture with price for a net of 10 kg.
-  no price/no mention of onions for pick yourself as the harvest season is over. 

edit : added example's pictures for potatoes
![P_20241204_093225](https://github.com/user-attachments/assets/08ba4198-e3db-40ab-b9b7-2473a1661f6c)
![P_20241204_093235](https://github.com/user-attachments/assets/f67e4dfc-16b3-478e-96c6-4a5654d7837d)

**Comment by raphodn** [comment]
### V1

List of discount types : 
- Quantity
- ~Promotion~ Sale
- Seasonal
- Loyalty program
- Expires soon
- Pick it yourself
- Other

ok for you ?

**Reply by serpico** [reply]
yeah, looks good, looking forward to trying it out.
Will it be "backward compatible" ? like users will have the ability the edit their prices to match the new discount types ?

**Reply by raphodn** [reply]
thanks for the feedback :pray: 
yes i'll make it editable :+1:

**Reply by raphodn** [reply]
v1 in prod :)

**Reply by raphodn** [reply]
need to add "second hand"

**Comment by julia-gx** [comment]
- Buy two, get the third (cheapest) for a cent (two different brands); no loyalty card required. 
- Mix & Match: buy one, get the second (cheaper) item 50% off (many brands); with a loyalty card. 
- With a loyalty card X% off, without loyalty card only Y% off 
- X% off for two products of different brands 
- Buy 2 the same brand (or more) for X.
- Buy 2 the same brand (or more) and get X% off for both.
- Buy 1 and get X% off for the second (the same brands).

**Comment by julia-gx** [comment]
<img width="853" height="893" alt="screen" src="https://github.com/user-attachments/assets/c03ed86b-0e77-41c3-a1f7-9b2f9f9e203f" />


Some concept draft for iteration.

**Comment by julia-gx** [comment]
Hello,
How can I apply: 
if you buy one item, you pay X per item.
if you buy three at once, you pay Y per item.
